--- dma.c	2017-12-06 08:32:37.000000000 +0500
+++ dma.c.bad	2017-12-06 11:20:18.068491273 +0500
@@ -163,10 +164,10 @@
     dev->regs->STREAM[stream].FCR = fifo_flags & 0x87; // mask out reserved bits
     dev->regs->STREAM[stream].CR = flags & 0x0feffffe; // mask out reserved and enable
 }
+#endif
 
-
-// compatible way
-void dma_init_transfer(dma_stream stream, DMA_InitTypeDef *v){
+// ST similar way
+void dma_init_transfer(dma_stream stream, DMA_InitType *v){
     uint32_t tmpreg;
 
     const dma_dev * dev=DMAS[(stream>>4) & 3]; 
@@ -199,11 +200,13 @@
   /* Set PL bits according to DMA_Priority value */
   /* Set MBURST bits according to DMA_MemoryBurst value */
   /* Set PBURST bits according to DMA_PeripheralBurst value */
-  tmpreg |= v->DMA_Channel | v->DMA_DIR |
+    tmpreg |= v->DMA_flags; 
+/*
+  v->DMA_Channel | v->DMA_DIR | v->DMA_Priority |
             v->DMA_PeripheralInc | v->DMA_MemoryInc |
-            v->DMA_PeripheralDataSize | v->DMA_MemoryDataSize |
-            v->DMA_Mode | v->DMA_Priority |
+            v->DMA_PeripheralDataSize | v->DMA_MemoryDataSize | v->DMA_Mode | 
             v->DMA_MemoryBurst | v->DMA_PeripheralBurst;
+*/         
 
   /* Write to DMAy Streamx CR register */
   DMAy_Streamx->CR = tmpreg;
@@ -218,7 +221,7 @@
   /* Configure DMAy Streamx FIFO: 
     Set DMDIS bits according to DMA_FIFOMode value 
     Set FTH bits according to DMA_FIFOThreshold value */
-  tmpreg |= v->DMA_FIFOMode | v->DMA_FIFOThreshold;
+    tmpreg |= v->DMA_FIFO_flags;
 
   /* Write to DMAy Streamx CR */
   DMAy_Streamx->FCR = tmpreg;
